File: Chap03/GoLang/lambda-closure.go
package main

import (
	"fmt"
)

func GetClosure() func() int {
	i := 0
	inc := func() int {
		i++
		return i
	}
	return inc
}


func main() {

	n := func(x, y int) int { return x + y }(3,2)
	fmt.Printf("calc %d\n", n)

	f := func(x, y int) int { return x + y }
	fmt.Printf("func %d\n", f(3,2))

	c := GetClosure()
	fmt.Printf("inc 1 -> %d\n", c())
	fmt.Printf("inc 2 -> %d\n", c())
	fmt.Printf("inc 3 -> %d\n", c())

}

File: Chap03/Python/lambda-closure.py
# -*- coding: utf-8 -*-

n = (lambda x,y: x+y)( 3, 3 )
print "calc = {}".format(n)

f = lambda x,y: x+y
print "func = {}".format(f(2,3))

def get_closure():
    def inc():
        inc.cnt += 1
        return inc.cnt
    inc.cnt = 0
    return inc


c = get_closure()

print "1 -> {}".format(c())
print "2 -> {}".format(c())
print "3 -> {}".format(c())
print "4 -> {}".format(c())


'''
 DO NOT WORK

def get_closure():
    cnt = 0
    def inc():
        cnt += 1
        return cnt
    return inc
'''


